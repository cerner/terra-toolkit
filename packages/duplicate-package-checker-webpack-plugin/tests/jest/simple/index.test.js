/**
 * @jest-environment node
 */

const webpack = require('webpack');
const assert = require('assert');
const chalk = require('chalk');
const MakeConfig = require('./make.webpack.config');

const chalkLevel = chalk.level;

describe('Simple dependency tree', () => {
  beforeEach(() => {
    /**
     * Travis-ci doesn't support colorized string generation with Chalk. To prevent snapshot mismatches
     * with Travis's test output, Chalk is disabled for snapshot tests.
     */
    chalk.level = 0;
  });

  afterEach(() => {
    chalk.level = chalkLevel;
  });

  it('should output warnings', (done) => {
    webpack(MakeConfig(), (err, stats) => {
      expect(stats.compilation.warnings[0].message).toMatchSnapshot();
      expect(stats.compilation.warnings[1].message).toMatchSnapshot();
      assert(stats.compilation.warnings.length === 2);
      done();
    });
  });

  it('should output warnings in verbose', (done) => {
    webpack(MakeConfig({ verbose: true }), (err, stats) => {
      expect(stats.compilation.warnings[0].message).toMatchSnapshot();
      expect(stats.compilation.warnings[1].message).toMatchSnapshot();
      assert(stats.compilation.warnings.length === 2);
      done();
    });
  });

  it('should output errors', (done) => {
    webpack(MakeConfig({ emitError: true }), (err, stats) => {
      expect(stats.compilation.errors[0].message).toMatchSnapshot();
      expect(stats.compilation.errors[1].message).toMatchSnapshot();
      assert(stats.compilation.errors.length === 2);
      assert(stats.compilation.warnings.length === 0);
      done();
    });
  });

  it('should output errors in production mode', (done) => {
    webpack(MakeConfig({ emitError: true }, 'production'), (
      err,
      stats,
    ) => {
      expect(stats.compilation.errors[0].message).toMatchSnapshot();
      expect(stats.compilation.errors[1].message).toMatchSnapshot();
      assert(stats.compilation.errors.length === 2);
      assert(stats.compilation.warnings.length === 0);
      done();
    });
  });

  it('should ignore excluded duplicates by name', (done) => {
    webpack(
      MakeConfig({
        exclude(instance) {
          return instance.name === 'a';
        },
      }),
      (err, stats) => {
        expect(stats.compilation.warnings[0].message).toMatchSnapshot();
        assert(stats.compilation.warnings.length === 1);
        done();
      },
    );
  });

  it('should ignore excluded duplicates by issuer', (done) => {
    webpack(
      MakeConfig({
        exclude(instance) {
          return instance.issuer === './entry.js';
        },
      }),
      (err, stats) => {
        assert(stats.compilation.warnings.length === 0);
        done();
      },
    );
  });
  it('should respect showHelp option', (done) => {
    webpack(
      MakeConfig({
        showHelp: false,
      }),
      (err, stats) => {
        expect(stats.compilation.warnings[0]).toMatchSnapshot();
        expect(stats.compilation.warnings[1]).toMatchSnapshot();
        done();
      },
    );
  });
  it('should output errors if duplicate packages are found within alwaysEmitErrorsFor', (done) => {
    webpack(MakeConfig({ alwaysEmitErrorsFor: ['a'] }), (err, stats) => {
      expect(stats.compilation.errors[0].message).toMatchSnapshot();
      expect(stats.compilation.warnings[0].message).toMatchSnapshot();
      assert(stats.compilation.errors.length === 1);
      assert(stats.compilation.warnings.length === 1);
      done();
    });
  });
});
