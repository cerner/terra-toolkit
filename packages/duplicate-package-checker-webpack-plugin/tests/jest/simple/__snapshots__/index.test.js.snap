// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple dependency tree should ignore excluded duplicates by name 1`] = `
"b
  Multiple versions of b found:
    1.0.0 ./~/b
    2.0.0 ./~/c/~/d/~/b

Check how you can resolve duplicate packages: 
https://github.com/cerner/terra-toolkit/tree/main/packages/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle
"
`;

exports[`Simple dependency tree should output errors 1`] = `
"a
  Multiple versions of a found:
    1.0.0 ./~/a
    2.0.0 ./~/b/~/a
"
`;

exports[`Simple dependency tree should output errors 2`] = `
"b
  Multiple versions of b found:
    1.0.0 ./~/b
    2.0.0 ./~/c/~/d/~/b

Check how you can resolve duplicate packages: 
https://github.com/cerner/terra-toolkit/tree/main/packages/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle
"
`;

exports[`Simple dependency tree should output errors if duplicate packages are found within alwaysEmitErrorsFor 1`] = `
"a
  Multiple versions of a found:
    1.0.0 ./~/a
    2.0.0 ./~/b/~/a
"
`;

exports[`Simple dependency tree should output errors if duplicate packages are found within alwaysEmitErrorsFor 2`] = `
"b
  Multiple versions of b found:
    1.0.0 ./~/b
    2.0.0 ./~/c/~/d/~/b

Check how you can resolve duplicate packages: 
https://github.com/cerner/terra-toolkit/tree/main/packages/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle
"
`;

exports[`Simple dependency tree should output errors in production mode 1`] = `
"a
  Multiple versions of a found:
    1.0.0 ./~/a
    2.0.0 ./~/b/~/a
"
`;

exports[`Simple dependency tree should output errors in production mode 2`] = `
"b
  Multiple versions of b found:
    1.0.0 ./~/b
    2.0.0 ./~/c/~/d/~/b

Check how you can resolve duplicate packages: 
https://github.com/cerner/terra-toolkit/tree/main/packages/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle
"
`;

exports[`Simple dependency tree should output warnings 1`] = `
"a
  Multiple versions of a found:
    1.0.0 ./~/a
    2.0.0 ./~/b/~/a
"
`;

exports[`Simple dependency tree should output warnings 2`] = `
"b
  Multiple versions of b found:
    1.0.0 ./~/b
    2.0.0 ./~/c/~/d/~/b

Check how you can resolve duplicate packages: 
https://github.com/cerner/terra-toolkit/tree/main/packages/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle
"
`;

exports[`Simple dependency tree should output warnings in verbose 1`] = `
"a
  Multiple versions of a found:
    1.0.0 ./~/a from ./entry.js
    2.0.0 ./~/b/~/a from ./~/b/index.js
"
`;

exports[`Simple dependency tree should output warnings in verbose 2`] = `
"b
  Multiple versions of b found:
    1.0.0 ./~/b from ./entry.js
    2.0.0 ./~/c/~/d/~/b from ./~/c/~/d/index.js

Check how you can resolve duplicate packages: 
https://github.com/cerner/terra-toolkit/tree/main/packages/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle
"
`;

exports[`Simple dependency tree should respect showHelp option 1`] = `
[Error: a
  Multiple versions of a found:
    1.0.0 ./~/a
    2.0.0 ./~/b/~/a
]
`;

exports[`Simple dependency tree should respect showHelp option 2`] = `
[Error: b
  Multiple versions of b found:
    1.0.0 ./~/b
    2.0.0 ./~/c/~/d/~/b
]
`;
